File: src-tauri/src/opencode_cfg.rs
use anyhow::Result;
use serde_json::json;

/// 生成 opencode2api 的 config.json
/// active_socks5 指向 sing-box 的 SOCKS5 端口
pub fn build_opencode_config(singbox_port: u16) -> Result<String> {
    let proxy_addr = format!("127.0.0.1:{}", singbox_port);
    let cfg = crate::config::Config::load().unwrap_or_default();
    let show_node_prefix = cfg.show_node_prefix.unwrap_or(false);
    let config = json!({
        "model_alias": {
            "deepseek-v4-flash": "deepseek-v4-flash-free",
            "mimo-v2.5": "mimo-v2.5-free",
            "ling-3.0-flash": "ling-3.0-flash-free",
            "nemotron-3-ultra": "nemotron-3-ultra-free",
            "north-mini-code": "north-mini-code-free",
            "laguna-s-2.1": "laguna-s-2.1-free"
        },
        "reasoning_effort_map": {
            "minimal": "low",
            "medium": "medium",
            "high": "high"
        },
        "force_disable_thinking": false,
        "socks5_proxies": [
            {
                "name": "singbox",
                "addr": proxy_addr,
                "username": "",
                "password": ""
            }
        ],
        "active_socks5": proxy_addr,
        "show_node_prefix": show_node_prefix
    });

    Ok(serde_json::to_string_pretty(&config)?)
}

/// 生成统一网关配置：池内实例作为 SOCKS5 代理列表，路由模式默认 failover
/// （成功不动游标、失败/限流/额度耗尽才切下一个健康实例）。
/// port_names 提供 singbox 端口 → 实例名映射，写入 socks5_proxies[].name，
/// 供 Go 侧流式输出显示「🤖 实例名 · 模型」（而非 SOCKS5 地址）。
pub fn build_opencode_router_config(
    singbox_ports: &[u16],
    port_names: &[(u16, String)],
    route_mode: &str,
) -> Result<String> {
    let name_for_port = |port: u16| -> String {
        port_names
            .iter()
            .find(|(p, _)| *p == port)
            .map(|(_, n)| n.clone())
            .unwrap_or_else(|| format!("instance-{}", port))
    };
    let proxies: Vec<serde_json::Value> = singbox_ports
        .iter()
        .map(|port| {
            json!({
                "name": name_for_port(*port),
                "addr": format!("127.0.0.1:{}", port),
                "username": "",
                "password": ""
            })
        })
        .collect();
    // 流内超时切换区间配置：从 manager 配置读取，未设置用默认值（毫秒）
    let cfg = crate::config::Config::load().unwrap_or_default();
    let ttft_min = cfg.timeout_ttft_min_ms.unwrap_or(10000);
    let ttft_max = cfg.timeout_ttft_max_ms.unwrap_or(10000);
    let silence_min = cfg.timeout_silence_min_ms.unwrap_or(5000);
    let silence_max = cfg.timeout_silence_max_ms.unwrap_or(5000);
    let probe_min = cfg.failover_probe_min.unwrap_or(2);
    let probe_max = cfg.failover_probe_max.unwrap_or(3);
    let call_log_max = cfg.call_log_max.unwrap_or(5000);
    let show_node_prefix = cfg.show_node_prefix.unwrap_or(false);
    let config = json!({
        "model_alias": {
            "deepseek-v4-flash": "deepseek-v4-flash-free",
            "mimo-v2.5": "mimo-v2.5-free",
            "ling-3.0-flash": "ling-3.0-flash-free",
            "nemotron-3-ultra": "nemotron-3-ultra-free",
            "north-mini-code": "north-mini-code-free",
            "laguna-s-2.1": "laguna-s-2.1-free"
        },
        "reasoning_effort_map": {
            "minimal": "low",
            "medium": "medium",
            "high": "high"
        },
        "force_disable_thinking": false,
        "socks5_proxies": proxies,
        "active_socks5": "__round_robin__",
        "route_mode": route_mode,
        "timeout_ttft_min_ms": ttft_min,
        "timeout_ttft_max_ms": ttft_max,
        "timeout_silence_min_ms": silence_min,
        "timeout_silence_max_ms": silence_max,
        "failover_probe_min": probe_min,
        "failover_probe_max": probe_max,
        "call_log_max": call_log_max,
        "show_node_prefix": show_node_prefix
    });

    Ok(serde_json::to_string_pretty(&config)?)
}

#[cfg(test)]
mod tests {
    use super::*;

    // 与 config 模块共用串行锁：本模块测试会 set OPCODE2API_DATA_DIR，
    // 而 Config::set()/load() 依赖该进程级 env 变量，需串行避免互相覆盖。
    fn lock() -> std::sync::MutexGuard<'static, ()> {
        crate::config::CONFIG_TEST_LOCK
            .lock()
            .unwrap_or_else(|e| e.into_inner())
    }

    #[test]
    fn test_build_opencode_config() {
        let _guard = lock();
        // 隔离数据目录：build_opencode_config 会读 Config::load()，
        // 避免受开发者本机 config.json 的 show_node_prefix 设置影响
        let test_dir = std::env::temp_dir().join("opencode2api-cfg-test");
        let _ = std::fs::remove_dir_all(&test_dir);
        unsafe { std::env::set_var("OPCODE2API_DATA_DIR", &test_dir) };
        let config = build_opencode_config(7890).unwrap();
        let v: serde_json::Value = serde_json::from_str(&config).unwrap();
        assert_eq!(v["active_socks5"], "127.0.0.1:7890");
        assert_eq!(v["socks5_proxies"][0]["addr"], "127.0.0.1:7890");
        assert!(v["model_alias"]["deepseek-v4-flash"].is_string());
        assert_eq!(v["force_disable_thinking"], false);
        // show_node_prefix 默认 false（默认关闭）
        assert_eq!(v["show_node_prefix"], false);
        unsafe { std::env::remove_var("OPCODE2API_DATA_DIR") };
    }

    #[test]
    fn test_build_opencode_config_different_port() {
        let _guard = lock();
        let config = build_opencode_config(7892).unwrap();
        let v: serde_json::Value = serde_json::from_str(&config).unwrap();
        assert_eq!(v["active_socks5"], "127.0.0.1:7892");
    }

    #[test]
    fn test_build_opencode_router_config() {
        let _guard = lock();
        // 隔离数据目录（同 test_build_opencode_config）
        let test_dir = std::env::temp_dir().join("opencode2api-cfg-test");
        let _ = std::fs::remove_dir_all(&test_dir);
        unsafe { std::env::set_var("OPCODE2API_DATA_DIR", &test_dir) };
        let names = vec![
            (18001u16, "日本1".to_string()),
            (18002u16, "美国2".to_string()),
        ];
        let config = build_opencode_router_config(&[18001, 18002], &names, "failover").unwrap();
        let v: serde_json::Value = serde_json::from_str(&config).unwrap();
        assert_eq!(v["active_socks5"], "__round_robin__");
        assert_eq!(v["route_mode"], "failover");
        assert_eq!(v["socks5_proxies"].as_array().unwrap().len(), 2);
        // name 应取真实实例名（而非占位 instance-N）
        assert_eq!(v["socks5_proxies"][0]["name"], "日本1");
        assert_eq!(v["socks5_proxies"][0]["addr"], "127.0.0.1:18001");
        assert_eq!(v["socks5_proxies"][1]["name"], "美国2");
        // show_node_prefix 默认 false
        assert_eq!(v["show_node_prefix"], false);
        unsafe { std::env::remove_var("OPCODE2API_DATA_DIR") };
    }

    #[test]
    fn test_build_opencode_router_config_round_robin() {
        let _guard = lock();
        let names = vec![
            (18001u16, "a".to_string()),
            (18002u16, "b".to_string()),
            (18003u16, "c".to_string()),
        ];
        let config =
            build_opencode_router_config(&[18001, 18002, 18003], &names, "round_robin").unwrap();
        let v: serde_json::Value = serde_json::from_str(&config).unwrap();
        assert_eq!(v["route_mode"], "round_robin");
        assert_eq!(v["socks5_proxies"].as_array().unwrap().len(), 3);
        assert_eq!(v["socks5_proxies"][2]["name"], "c");
    }
}

File: core/manager/opencodecfg.go
// 实例/网关 opencode2api.json 配置生成（Rust opencode_cfg.rs 移植）。
package manager

import (
	"encoding/json"
	"fmt"
	"net"
	"sort"
	"strconv"
	"strings"
)

// defaultFreeAliases 免费模型别名（-free 后缀隐藏）。
var defaultFreeAliases = []struct{ from, to string }{
	{"deepseek-v4-flash", "deepseek-v4-flash-free"},
	{"mimo-v2.5", "mimo-v2.5-free"},
	{"ling-3.0-flash", "ling-3.0-flash-free"},
	{"nemotron-3-ultra", "nemotron-3-ultra-free"},
	{"north-mini-code", "north-mini-code-free"},
	{"laguna-s-2.1", "laguna-s-2.1-free"},
}

func aliasMap() map[string]string {
	m := map[string]string{}
	for _, a := range defaultFreeAliases {
		m[a.from] = a.to
	}
	return m
}

func reasoningMap() map[string]string {
	return map[string]string{"minimal": "low", "medium": "medium", "high": "high"}
}

// upstreamProxyAddr 把配置的 upstream_proxy 归一化为裸 host:port。
// socks.go 的 socks5Dial 直接 net.Dial("tcp", addr)，地址不接受带 scheme 的前缀——
// 因此 socks5:// 与 http:// 前缀在此剥离（Clash mixed-port 同时支持两种协议，统一
// 存 host:port）。空串 / 无法解析出 host:port（如无端口）返回 ""（回退直连 = 现状）。
func upstreamProxyAddr(raw string) string {
	addr := strings.TrimSpace(raw)
	if i := strings.Index(addr, "://"); i >= 0 {
		addr = addr[i+len("://"):]
	}
	if addr == "" {
		return ""
	}
	if _, port, err := net.SplitHostPort(addr); err != nil {
		return ""
	} else if _, err := strconv.ParseUint(port, 10, 16); err != nil {
		// SplitHostPort 不校验端口数字（如 "7897/path" 也能拆出），
		// socks5Dial 直接 net.Dial 需要数字端口。
		return ""
	}
	return addr
}

// buildOpenCodeCfg 单实例配置：默认本地 sing-box 作为唯一 SOCKS5 出口；
// 配置了上游代理（E1）时出口直接指向该代理（跳过 sing-box）。
func (m *Manager) buildOpenCodeCfg(singboxPort uint16) ([]byte, error) {
	appCfg := m.loadConfig()
	proxyAddr := upstreamProxyAddr(appCfg.UpstreamProxy)
	proxyName := "singbox"
	if proxyAddr == "" {
		// 未配置 / 非法值：保持现状——本地 sing-box 出口。
		proxyAddr = fmt.Sprintf("127.0.0.1:%d", singboxPort)
	} else {
		proxyName = "upstream"
	}
	cfg := map[string]any{
		"model_alias":            aliasMap(),
		"reasoning_effort_map":   reasoningMap(),
		"force_disable_thinking": false,
		"socks5_proxies": []any{map[string]any{
			"name": proxyName, "addr": proxyAddr,
		}},
		"active_socks5":    proxyAddr,
		"show_node_prefix": appCfg.ShowNodePrefix,
	}
	// 透传厂商注册表 + 路由：实例子进程与核心一致，能注册多厂商（如 windsurf）
	cfg = injectVendorConfig(cfg, appCfg)
	return json.MarshalIndent(cfg, "", "  ")
}

// BuildOpenCodeCfgFor 导出包装（main 装配实例接缝用）。
func (m *Manager) BuildOpenCodeCfgFor(port uint16) ([]byte, error) { return m.buildOpenCodeCfg(port) }

// buildRouterCfg 统一网关路由配置：多 sing-box 出口 + route_mode + 超时区间。
func (m *Manager) buildRouterCfg(singboxPorts []uint16, portNames map[uint16]string, routeMode string) ([]byte, error) {
	appCfg := m.loadConfig()
	proxies := []any{}
	sorted := append([]uint16(nil), singboxPorts...)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i] < sorted[j] })
	for _, p := range sorted {
		name := portNames[p]
		if name == "" {
			name = fmt.Sprintf("instance-%d", p)
		}
		proxies = append(proxies, map[string]any{
			"name": name, "addr": fmt.Sprintf("127.0.0.1:%d", p),
		})
	}
	if routeMode == "" {
		routeMode = "round_robin"
	}
	cfg := map[string]any{
		"model_alias":            aliasMap(),
		"reasoning_effort_map":   reasoningMap(),
		"force_disable_thinking": false,
		"socks5_proxies":         proxies,
		"active_socks5":          "__round_robin__",
		"route_mode":             routeMode,
		"show_node_prefix":       appCfg.ShowNodePrefix,
	}
	// 超时/日志区间（0 = 用 core 默认）
	applyIf := func(key string, v int64) {
		if v > 0 {
			cfg[key] = v
		}
	}
	applyIf("timeout_ttft_min_ms", appCfg.TimeoutTTFTMinMS)
	applyIf("timeout_ttft_max_ms", appCfg.TimeoutTTFTMaxMS)
	applyIf("timeout_silence_min_ms", appCfg.TimeoutSilenceMinMS)
	applyIf("timeout_silence_max_ms", appCfg.TimeoutSilenceMaxMS)
	applyIf("failover_probe_min", appCfg.FailoverProbeMin)
	applyIf("failover_probe_max", appCfg.FailoverProbeMax)
	applyIf("call_log_max", appCfg.CallLogMax)
	// P2 性能模式：熔断阈值 / 半开间隔 / 开关透传（>0 才写，未配置子进程用默认）。
	applyIf("pool_breaker_threshold", int64(appCfg.PoolBreakerThreshold))
	applyIf("pool_halfopen_interval_sec", int64(appCfg.PoolHalfOpenIntervalSec))
	if appCfg.PoolPerformanceMode != nil {
		cfg["pool_performance_mode"] = *appCfg.PoolPerformanceMode
	}
	// P2b 请求级竞速：并行扇出数（>0 才写，未配置子进程用默认 2）。
	applyIf("pool_race_copies", int64(appCfg.PoolRaceCopies))
	// S1 竞速整体预算（毫秒；>0 才写，未配置子进程用默认 10000）。
	applyIf("race_budget_ms", int64(appCfg.RaceBudgetMS))
	// 透传厂商注册表 + 路由：网关子进程与核心一致，能注册多厂商（如 windsurf）
	cfg = injectVendorConfig(cfg, appCfg)
	return json.MarshalIndent(cfg, "", "  ")
}

// injectVendorConfig 把管理器配置里的厂商注册表与路由写入子进程配置。
// 子进程（实例子进程 / 网关子进程）与核心同二进制，读同一格式的 opencode2api.json，
// 不带 providers 则只认 opencode —— 多厂商（windsurf 等）在 exe 实例/网关形态下不可用。
func injectVendorConfig(cfg map[string]any, appCfg Config) map[string]any {
	if len(appCfg.Providers) > 0 {
		cfg["providers"] = appCfg.Providers
	}
	if len(appCfg.Routing) > 0 {
		cfg["routing"] = appCfg.Routing
	}
	return cfg
}
